relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -0,0 +1,766 @@
+    #[serde(default)]
+    pub project_ids: Vec<ProjectId>,
+    /// A set of Glob patterns for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub releases: GlobPatterns,
+    /// A set of user_segments for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub user_segments: Vec<LowerCaseString>,
+    #[serde(default)]
+    pub environments: Vec<LowerCaseString>,
+    /// the sampling rate for trace matching this rule
+    pub sample_rate: f64,
+}
+
+#[derive(Debug, Clone, Serialize, Default)]
+pub struct LowerCaseString(String);
+
+impl LowerCaseString {
+    pub fn new(val: &str) -> LowerCaseString {
+        LowerCaseString(val.to_lowercase())
+    }
+}
+
+impl From<&str> for LowerCaseString {
+    fn from(val: &str) -> Self {
+        LowerCaseString(val.to_lowercase())
+    }
+}
+
+impl PartialEq for LowerCaseString {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Eq for LowerCaseString {}
+
+struct LowerCaseStringVisitor;
+
+impl<'de> Visitor<'de> for LowerCaseStringVisitor {
+    type Value = LowerCaseString;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a string to be converted to a lowercase string ")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E> {
+        Ok(LowerCaseString::new(v))
+    }
+}
+
+impl<'de> Deserialize<'de> for LowerCaseString {
+    fn deserialize<D>(deserializer: D) -> Result<LowerCaseString, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_string(LowerCaseStringVisitor)
+    }
+}
+
+impl SamplingRule {
+    /// Tests whether a rule matches a trace context
+    fn matches(
+        &self,
+        release: &Option<String>,
+        user_segment: &Option<LowerCaseString>,
+        environment: &Option<LowerCaseString>,
+        project_id: ProjectId,
+    ) -> bool {
+        // match against the project
+        if !self.project_ids.is_empty() && !self.project_ids.contains(&project_id) {
+            return false;
+        }
+        // match against the release
+        if !self.releases.is_empty() {
+            match release {
+                None => return false,
+                Some(ref release) => {
+                    if !self.releases.is_match(release) {
+                        return false;
+                    }
+                }
+            }
+        }
+        // match against the user_segment
+        if !self.user_segments.is_empty() {
+            match user_segment {
+                None => return false,
+                Some(ref user_segment) => {
+                    if !self.user_segments.contains(user_segment) {
+                        return false;
+                    }
+                }
+            }
+        }
+        // match against the environment
+        if !self.environments.is_empty() {
```

Comment:
_nit_: Environments are lowest cardinality and cheapest to check, so I would start with this.